pupil_src/shared_modules/calibration_routines/data_processing.py

```diff
@@ -0,0 +1,200 @@
+import numpy as np
+
+# logging
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def filter_confidence(pupil_list, threshold):
+
+    len_pre_filter = len(pupil_list)
+
+    pupil_list = [p for p in pupil_list if p["confidence"] >= threshold]
+
+    len_post_filter = len(pupil_list)
+    try:
+        dismissed_percentage = 100 * (1.0 - len_post_filter / len_pre_filter)
+    except ZeroDivisionError:
+        pass  # empty pupil_list, is being handled in match_data
+    else:
+        logger.info(
+            "Dismissing {:.2f}% pupil data due to confidence < {:.2f}".format(
+                dismissed_percentage, threshold
+            )
+        )
+
+    return pupil_list
+
+
+def find_nearest_idx(array, value):
+    idx = np.searchsorted(array, value, side="left")
+    try:
+        if abs(value - array[idx - 1]) < abs(value - array[idx]):
+            return idx - 1
+        else:
+            return idx
+    except IndexError:
+        return idx - 1
+
+
+def match_data(g_pool, pupil_list, ref_list):
+    """
+    Returns binocular and monocular matched pupil datums and ref points. Uses a dispersion criterion to
+    dismiss matches which are too far apart.
+    """
+    if pupil_list and ref_list:
+        pass
+    else:
+        not_enough_data_error_msg = (
+            "Not enough ref point or pupil data available for calibration."
+        )
+        logger.error(not_enough_data_error_msg)
+        return {
+            "subject": "calibration.failed",
+            "reason": not_enough_data_error_msg,
+            "timestamp": g_pool.get_timestamp(),
+            "record": True,
+        }
+
+    pupil0 = [p for p in pupil_list if p["id"] == 0]
+    pupil1 = [p for p in pupil_list if p["id"] == 1]
+
+    matched_binocular_data = closest_matches_binocular(ref_list, pupil_list)
+    matched_pupil0_data = closest_matches_monocular(ref_list, pupil0)
+    matched_pupil1_data = closest_matches_monocular(ref_list, pupil1)
+
+    if len(matched_pupil0_data) > len(matched_pupil1_data):
+        matched_monocular_data = matched_pupil0_data
+    else:
+        matched_monocular_data = matched_pupil1_data
+
+    logger.info(
+        "Collected {} monocular calibration data.".format(len(matched_monocular_data))
+    )
+    logger.info(
+        "Collected {} binocular calibration data.".format(len(matched_binocular_data))
+    )
+    return (
+        matched_binocular_data,
+        matched_monocular_data,
+        matched_pupil0_data,
+        matched_pupil1_data,
+        pupil0,
+        pupil1,
+    )
+
+
+def closest_matches_binocular(ref_pts, pupil_pts, max_dispersion=1 / 15.0):
+    """
+    get pupil positions closest in time to ref points.
+    return list of dict with matching ref, pupil0 and pupil1 data triplets.
+    """
+    pupil0 = [p for p in pupil_pts if p["id"] == 0]
```

Comment:
This split has already been computed in `match_data`. You could user `pupil0` and `pupil1` as arguments already.